File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-5.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int n = 0;
        int n_1 = cost[cost.length-1];
        for(int i=cost.length-2; i>=0; i--){
            int t = cost[i] + Math.min(n, n_1);
            n = n_1;
            n_1 = t;
        }
        return Math.min(n_1, n);
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-5.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = -int(1e9)
        cur_min = 1
        cur_max = 1
        for num in nums:
            tmp = cur_min
            cur_min = min(num, num*cur_min, num*cur_max)
            cur_max = max(num, num*tmp, num*cur_max)
            res = max(res, cur_max)
        
        return res